irc/server.go

```diff
@@ -432,33 +432,142 @@ func (client *Client) getWhoisOf(target *Client, rb *ResponseBuffer) {
 	}
 }
 
-// rplWhoReply returns the WHO reply between one user and another channel/user.
-// <channel> <user> <host> <server> <nick> ( "H" / "G" ) ["*"] [ ( "@" / "+" ) ]
-// :<hopcount> <real name>
-func (client *Client) rplWhoReply(channel *Channel, target *Client, rb *ResponseBuffer) {
-	channelName := "*"
-	flags := ""
+type WhoFields uint
+
+const (
+	FieldType WhoFields = iota << 0
+	FieldChannel
+	FieldUsername
+	FieldIP
+	FieldHostname
+	FieldServer
+	FieldNickname
+	FieldFlags
+	FieldHops
+	FieldIdle
+	FieldAccount
+	FieldRank
+	FieldRealname
+)
 
-	if target.Away() {
-		flags = "G"
-	} else {
-		flags = "H"
+var whoFieldMap = map[rune]WhoFields{
```

Comment:
Here's what I'd actually do instead of defining these:

```go
type WhoFields [128]bool

func (wf *WhoFields) Set(field rune) {
	wf[int(field)] = true
}

func (wf *WhoFields) Has(field rune) bool {
	return wf[int(field)]
}
```

i.e., just use the codepoint value of 't' or 'a' to index directly into a fixed-size array of bools. (We use similar techniques for modes.)

Reply:
This technique would also need bounds checking (or else you could up it to `[256]bool` and operate on bytes instead of runes) but it's still probably worth it.